lib/charms/tempo_k8s/v1/charm_tracing.py

```diff
@@ -232,8 +244,18 @@ class UntraceableObjectError(TracingError):
     """Raised when an object you're attempting to instrument cannot be autoinstrumented."""
 
 
-def _get_tracing_endpoint(tracing_endpoint_getter, self, charm):
-    if isinstance(tracing_endpoint_getter, property):
+class TLSError(TracingError):
+    """Raised when the tracing endpoint is https but we don't have a cert yet."""
+
+
+def _get_tracing_endpoint(
+    tracing_endpoint_getter: Union[_GetterType, _InstanceAttribute],
+    self: ops.CharmBase,
+    charm: Type[ops.CharmBase],
+):
+    if isinstance(tracing_endpoint_getter, _InstanceAttribute):
+        tracing_endpoint = getattr(self, tracing_endpoint_getter)
+    elif isinstance(tracing_endpoint_getter, property):
```

Comment:
(for some reason I thought class attributes were also a possibility, so I wrote this with them in mind too.  But iiuc they're not, and you can ignore that part :) )

Can you get rid of `_InstanceAttribute` by doing a duck-typing approach?  Here's a demo of what I'm thinking:

```python
def get_val(obj, attr):
    """Returns the value of an attribute in an object that could be a method, property, instance variable, or class variable."""
    value_getter = getattr(obj, attr)
    try:
        value = value_getter()
        print("Callable - this was a method!")
    except TypeError:
        print("Not callable, so its a value or property that rendered itself to a value")
        value = value_getter
    return value

class Test:
    class_attribute = "class attribute"
    class_attribute_we_will_mask = "class attribute not yet masked"
    class_attribute_we_will_change_in_class = "class attribute not yet changed in class"

    def __init__(self, masked_class_attribute=None):
        self.instance_attribute = "instance attribute"
        if masked_class_attribute:
            self.class_attribute_we_will_mask = masked_class_attribute

    @property
    def my_property(self):
        return "my property"

    def my_method(self):
        return "my method"

t_unmasked = Test()
t_masked = Test("class attribute masked")

def print_results(x):
    print("___________________")
    print(f"{get_val(x, 'my_property')=}")
    print(f"{get_val(x, 'class_attribute')=}")
    print(f"{get_val(x, 'class_attribute_we_will_mask')=}")
    print(f"{get_val(x, 'instance_attribute')=}")
    print(f"{get_val(x, 'my_method')=}")

print("t_unmasked")
print_results(t_unmasked)

# OUTPUT:
# t_unmasked
# ___________________
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'my_property')='my property'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'class_attribute')='class attribute'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'class_attribute_we_will_mask')='class attribute not yet masked'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'instance_attribute')='instance attribute'
# Callable - this was a method!
# get_val(x, 'my_method')='my method'

print("t_masked")
print_results(t_masked)

# OUTPUT
# t_masked
# ___________________
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'my_property')='my property'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'class_attribute')='class attribute'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'class_attribute_we_will_mask')='class attribute masked'
# Not callable, so its a value or property that rendered itself to a value
# get_val(x, 'instance_attribute')='instance attribute'
# Callable - this was a method!
# get_val(x, 'my_method')='my method'


```

Reply:
this helps! I just have to delay the getattr call to simplify the current approach quite a lot.